src/oop/object_interactions/medium/q2/MainTest.java

```diff
@@ -0,0 +1,224 @@
+    @Tag("dependent1")
+    public void correctHearSecretMethod(String name, String secret, String[] allFriends, String friend, String friendSecret) throws Throwable {
+        Object[][] arguments = {
+                {name, String.class},
+                {secret, String.class},
+                {allFriends, String[].class}
+        };
+        Object[][] methodArguments = {
+                {friend, String.class},
+                {friendSecret, String.class}
+        };
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}),
+                "Your " + hearSecretMethodName + " method does not have the correct name or arguments.");
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}, Void.TYPE),
+                "Your " + hearSecretMethodName + " method does not have the correct return type.");
+        assertTrue(testClass.hasMethod(hearSecretMethodName, new Class<?>[]{String.class, String.class}, Void.TYPE, new String[]{"public"}),
+                "Your " + hearSecretMethodName + " method does not have the correct modifier.");
+        String[] ans = new String[allFriends.length];
+        for (int i = 0; i < allFriends.length; i++) {
+            if (friend.equals(allFriends[i]))
+                ans[i] = friendSecret;
+        }
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.callMethod(hearSecretMethodName, methodArguments, new String[]{"public"}, classInstance);
+        _assertArrayEquals(ans, testClass.getFieldValue(classInstance, friendSecretsAttributeName),
+                "Your " + hearSecretMethodName + " method does not change the " + friendSecretsAttributeName + " attribute correctly.");
+    }
+
+    @Test
+    public void writeInDiaryIsDefinedCorrectly() {
+        String incorrectMethodDefinition = "The %s method in the %s class is not defined correctly. Make sure it is declared, spelt correctly, and has the correct parameters.";
+        String incorrectModifierMessage = "The %s method in the %s class has the wrong visibility modifier.";
+        String incorrectReturnType = "The %s method in the %s class has the incorrect return type.";
+        assertTrue(testClass.hasMethod(writeInDiaryMethodName, null), String.format(incorrectMethodDefinition, writeInDiaryMethodName, className));
+        assertTrue(testClass.hasModifier(writeInDiaryMethodName, null, "public"), String.format(incorrectModifierMessage, writeInDiaryMethodName, className));
+        assertTrue(testClass.hasReturnType(writeInDiaryMethodName, null, Void.TYPE), String.format(incorrectReturnType, writeInDiaryMethodName, className));
+    }
+
+    @ParameterizedTest
+    @MethodSource("hearSecretInputProvider")
+    @Tag("dependent1")
+    public void correctWriteInDiaryMethod(String name, String secret, String[] allFriends, String friend, String friendSecret) throws Throwable {
```

Comment:
Please refactor this test so we can test if the method works properly if they have multiple secrets.